heat/core/dndarray.py

```diff
@@ -653,6 +653,84 @@ def fill_diagonal(self, value: float) -> DNDarray:
+            advanced_indexing = True
+            key = factories.array(key)
+            # DOES NOT WORK FOR SEQUENCE OF TENSORS OR DNDARRAYS, works for sequence of ndarrays though
+            # TODO: get outshape + outsplit; depends on wether key is bool or int and key.ndim
+        elif isinstance(key, tuple):
+            key = list(key)
+            for i, k in enumerate(key):
+                if isinstance(k, Iterable) or isinstance(key, DNDarray):
+                    advanced_indexing = True
+                    key[i] = factories.array(key[i])
+                    # DOES NOT WORK FOR SEQUENCE OF TENSORS OR DNDARRAYS, works for seq of ndarrays though
+                    # TODO: check for key.ndim = 0 and treat that as int
+                    # TODO: get outshape + outsplit; depends on wether key is bool or int and key.ndim
+            add_dims = sum(k is None for k in key)  # (np.newaxis is None)===true
+            ellipsis = sum(isinstance(k, type(...)) for k in key)
+            if ellipsis > 1:
+                raise ValueError("key can only contain 1 ellipsis")
+            elif ellipsis == 1:
+                expand_key = [slice(None)] * (arr.ndim + add_dims)
+                ellipsis_index = key.index(...)
+                expand_key[:ellipsis_index] = key[:ellipsis_index]
+                expand_key[ellipsis_index - len(key) :] = key[ellipsis_index + 1 :]
+                key = expand_key
+            if add_dims:
+                for i, k in reversed(enumerate(key)):
+                    if k is None:
+                        key[i] = slice(None)
+                        arr = arr.expand_dims(i - add_dims + 1)  # is the -1 correct?
+                        add_dims -= 1
+            # expand key to match the number of dimensions of the DNDarray
+            key = tuple(key + [slice(None)] * (arr.ndim - len(key)))
+        else:  # key is integer or slice
+            key = tuple([key] + [slice(None)] * (arr.ndim - 1))
+        return advanced_indexing, arr, key
+
+    def __get_local_slice(self, key: slice):
+        split = self.split
+        if split is None:
+            return key
+        key = stride_tricks.sanitize_slice(key, self.shape[split])
+        start, stop, step = key.start, key.stop, key.step
+        if step < 0:  # NOT supported by torch, should be filtered by torch_proxy
+            key = self.__get_local_slice(slice(stop + 1, start + 1, abs(step)))
+            if key is None:
+                return None
+            start, stop, step = key.start, key.stop, key.step
+            return slice(key.stop - 1, key.start - 1, -1 * key.step)
+
+        _, offsets = self.counts_displs()
+        offset = offsets[self.comm.rank]
+        range_proxy = range(self.lshape[split])
+        local_inds = range_proxy[start - offset : stop - offset]  # only works if stop - offset > 0
+        local_inds = local_inds[max(offset - start, 0) % step :: step]
```

Comment:
How about:
```
local_start = max(start-offset, 0)
local_stop = max(min(stop-offset, range_proxy.stop), 0)
local_inds = range_proxy[local_start:local_stop]
local_inds = local_inds[max(offset - start, 0) % step :: step]
```
?

Reply:
Good point.
So what you're doing by handling the negative values by hand is actually everything indexing the `range_proxy` would do, right? So there is no reason to sidestep into using a `range` (except the comfort of using `len(local_inds)` to find out if the local slice is empty) and we can directly use 
```python
local_slice = slice(local_start + max(offset - start, 0) % step, local_stop, step)
```